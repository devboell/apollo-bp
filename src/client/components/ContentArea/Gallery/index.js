import React from 'react'
import PropTypes from 'prop-types'

import GalleryData from 'containers/GalleryData'
import GenreContainer from 'containers/GenreContainer'

const Gallery = ({ match }) =>
  <GalleryData {...{ match }}>
    <GenreContainer />
  </GalleryData>

Gallery.propTypes = {
  match: PropTypes.shape({}).isRequired,
}

export default Gallery
