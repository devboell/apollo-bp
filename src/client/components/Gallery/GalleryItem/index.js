import React from 'react'
import PropTypes from 'prop-types'
import Title from './Title'
import Wrapper from './Wrapper'
import ThumbContainer from './ThumbContainer'
import ThumbLink from './ThumbLink'

const GalleryItem = ({ artistPath, painting }) =>
  <Wrapper>
    <ThumbContainer>
      <ThumbLink
        artistPath={artistPath}
        paintingName={painting.name}
      />
    </ThumbContainer>
    <Title>
      {painting.title}
    </Title>
  </Wrapper>

GalleryItem.propTypes = {
  artistPath: PropTypes.string.isRequired,
  painting: PropTypes.shape({
    title: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
}


export default GalleryItem
