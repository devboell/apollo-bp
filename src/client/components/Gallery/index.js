import React from 'react'
import PropTypes from 'prop-types'

const Gallery = ({ match }) =>
  <article>{match.params.artistName}</article>

Gallery.propTypes = {
  match: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
}
export default Gallery
