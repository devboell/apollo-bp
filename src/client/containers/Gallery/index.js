import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { withRouter } from 'react-router-dom'
import { gql, graphql } from 'react-apollo'
import { uniq, compose, map as _map } from 'lodash/fp'

import GenreChooser from 'components/GenreChooser'
import GalleryView from 'components/GalleryView'

import { setSelectedGenres, selectGenre, deselectGenre } from './actions'
import Wrapper from './Wrapper'

const artistGenres =
  compose(
    uniq,
    _map('genre'),
  )

export class Gallery extends React.Component {
  componentWillReceiveProps(nextProps) {
    const { artist, onSetSelectedGenres } = this.props
    if (nextProps.artist !== artist) {
      onSetSelectedGenres(artistGenres(nextProps.artist.paintings))
    }
  }

  render() {
    const {
      artist,
      loading,
      selectedGenres,
      onSelectGenre,
      onDeselectGenre,
    } = this.props

    const genreFilter = painting => selectedGenres.includes(painting.genre)

    return (
      loading
        ? <div>loading...</div>
        :
        <Wrapper>
          <GenreChooser
            genres={artistGenres(artist.paintings)}
            {...{
              selectedGenres,
              onSelectGenre,
              onDeselectGenre,
            }}
          />
          <GalleryView
            artistPath={artist.path}
            paintings={artist.paintings.filter(genreFilter)}
          />
        </Wrapper>
    )
  }
}

Gallery.propTypes = {
  loading: PropTypes.bool.isRequired,
  artist: PropTypes.shape({
    path: PropTypes.string,
    paintings: PropTypes.array,
  }),
  selectedGenres: PropTypes.arrayOf(PropTypes.string).isRequired,
  onSetSelectedGenres: PropTypes.func.isRequired,
  onSelectGenre: PropTypes.func.isRequired,
  onDeselectGenre: PropTypes.func.isRequired,
}

Gallery.defaultProps = {
  artist: {},
}

Gallery.displayName = 'Gallery'

export const ARTIST_BY_PATH_QUERY = gql`
  query ArtistByPathQuery($path: String) {
    artistByPath(path: $path) {
      path
      paintings {
        title
        name
        genre
      }
    }
  }
`

export const withArtist = graphql(ARTIST_BY_PATH_QUERY,
  {
    options: ({ match }) => ({ variables: { path: match.params.artistPath } }),
    props: ({ data: { artistByPath, loading } }) => ({
      artist: artistByPath,
      loading,
    }),
  })


export const mapStateToProps = state => ({
  selectedGenres: state.gallery.selectedGenres,
})

export const mapDispatchToProps = dispatch => ({
  onSetSelectedGenres: genres => dispatch(setSelectedGenres(genres)),
  onSelectGenre: genre => dispatch(selectGenre(genre)),
  onDeselectGenre: genre => dispatch(deselectGenre(genre)),
})

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  withRouter,
  withArtist,
)(Gallery)
