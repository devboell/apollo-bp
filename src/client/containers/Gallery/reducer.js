import types from './actionTypes'

const initialState = {
  selectedGenres: [],
}

export default (state = initialState, action) => {
  switch (action.type) {
    case types.SET_SELECTED_GENRES:
      return { ...state, selectedGenres: action.genres }
    case types.SELECT_GENRE:
      return { ...state, selectedGenres: [...state.selectedGenres, action.genre] }
    case types.DESELECT_GENRE:
      return {
        ...state,
        selectedGenres: state.selectedGenres.filter(g => g !== action.genre) }
    default:
      return state
  }
}
