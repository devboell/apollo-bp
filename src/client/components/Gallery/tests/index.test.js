import React from 'react'
import Gallery from '../index'

const match = {
  params: {
    artistName: 'Heroshige',
  },
}

it('Gallery component snapshot', () => {
  const wrapper = shallow(
    <Gallery {...{ match }} />,
  )
  expect(wrapper).toMatchSnapshot()
})
