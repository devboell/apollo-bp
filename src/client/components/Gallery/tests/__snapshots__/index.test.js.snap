// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gallery component snapshot 1`] = `
<article>
  Heroshige
</article>
`;
